whois-nrtm/src/main/java/net/ripe/db/whois/nrtm/NrtmServer.java

```diff
@@ -50,33 +52,39 @@ public void start() {
             return;
         }
 
-        serverChannel = bootstrapChannel(nrtmServerPipelineFactory, nrtmPort, "NRTM DUMMIFIER");
-        port = ((InetSocketAddress) serverChannel.getLocalAddress()).getPort();
+        serverChannel = bootstrapChannel(nrtmServerChannelInitializer, nrtmPort);
+        port = ((InetSocketAddress) serverChannel.localAddress()).getPort();
+        LOGGER.info("NRTM server listening on port {}", port);
     }
 
-    private Channel bootstrapChannel(final ChannelPipelineFactory serverPipelineFactory, final int port, final String instanceName) {
-        channelFactory = new NioServerSocketChannelFactory();
-        final ServerBootstrap bootstrap = new ServerBootstrap(channelFactory);
-        bootstrap.setPipelineFactory(serverPipelineFactory);
-        bootstrap.setOption("backlog", 200);
-        // apply TCP options to accepted Channels. Ref. https://netty.io/3.10/guide/
-        bootstrap.setOption("child.tcpNoDelay", true);
-        bootstrap.setOption("child.keepAlive", true);
-
-        final Channel channel = bootstrap.bind(new InetSocketAddress(port));
-        final int actualPort = ((InetSocketAddress) channel.getLocalAddress()).getPort();
-        LOGGER.info("NRTM server listening on port {} ({})", actualPort, instanceName);
-        return channel;
+    // TODO: [ES] channel was not returned?
+    private Channel bootstrapChannel(final NrtmServerChannelInitializer serverChannelInitializer, final int nrtmPort) {
+        bossGroup = new NioEventLoopGroup();
+        workerGroup = new NioEventLoopGroup();
+        final ServerBootstrap bootstrap = new ServerBootstrap();
+        bootstrap.group(bossGroup, workerGroup)
+                .channel(NioServerSocketChannel.class)
+                .childHandler(serverChannelInitializer)
+                .option(ChannelOption.SO_BACKLOG, 200)
+                .childOption(ChannelOption.TCP_NODELAY, true)
+                .childOption(ChannelOption.SO_KEEPALIVE, true);
+
+        try {
+            final ChannelFuture channelFuture = bootstrap.bind(new InetSocketAddress(nrtmPort)).sync();
+            return channelFuture.channel();
+        } catch (InterruptedException e) {
+            throw new IllegalStateException("NRTM server start up failed", e);
+        }
     }
 
     @Override
     public void stop(final boolean force) {
+        workerGroup.shutdownGracefully();
+        bossGroup.shutdownGracefully();
         if (nrtmEnabled) {
             if (force) {
                 LOGGER.info("Shutting down");
-                if (channelFactory != null) {
-                    channelFactory.shutdown();
-                }
+                // TODO: [ES] do we need to shutdown bossGroup and workerGroup?
```

Comment:
That seems to be the common/best practice, judging by examples from the netty project [here](https://github.com/netty/netty/tree/4.1/example/src/main/java/io/netty/example) only difference the shutdown happens in the `finally` block

Reply:
@eshryane tagging you in case you missed my reply. Ok with the response? Or your question was about something else?

Reply:
Thanks I removed the comment